flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/WithinCheckpointFileMergingSnapshotManager.java

```diff
@@ -50,6 +50,43 @@ public WithinCheckpointFileMergingSnapshotManager(String id, Executor ioExecutor
         writablePhysicalFilePool = new HashMap<>();
     }
 
+    // ------------------------------------------------------------------------
+    //  CheckpointListener
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void notifyCheckpointComplete(SubtaskKey subtaskKey, long checkpointId)
```

Comment:
Why `notifyCheckpointSubsumed` is not overriden? 

IIUC, checkpoint subsume and complete are triggered at the same time([code](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java#L1082-L1085)),  although the effect of the two methods is the same, it feels more straightforward to close the physical file in `notifyCheckpointSubsumed`, WDYT?

Reply:
When a checkpoint is subsumed, it must be notified completed once before. So the ```removeAndCloseFiles``` has been called for this checkpoint when it is subsumed, thus no need another call for it.

And for notification loss scenarios, we will implement ```discardCheckpoint``` to leverage the watermark mechanism to do something like ```removeAndCloseFiles``` for each SubtaskKey to close the files in FLINK-32076.